realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,271 @@
+            } else if (cls == DynamicRealm.class) {
+                return DYNAMIC_REALM;
+            }
+
+            throw new IllegalArgumentException("The type of Realm class must be Realm or DynamicRealm.");
+        }
+    }
+    // Separated references and counters for typed Realm and dynamic Realm.
+    private final EnumMap<RealmCacheType, RefAndCount> refAndCountMap;
+
+    private RealmConfiguration configuration;
+    // Column indices are cached to speed up opening typed Realm. If a Realm instance is created in one thread, creating
+    // Realm instances in other threads doesn't have to initialize the column indices again.
+    private ColumnIndices typedColumnIndices;
+
+    // Different Realm configuration with same path are not allowed.
+    private static Map<String, RealmCache> cachesMap = new HashMap<>();
+
+    private static final String DIFFERENT_KEY_MESSAGE = "Wrong key used to decrypt Realm.";
+
+    private RealmCache(RealmConfiguration config) {
+        configuration = config;
+        refAndCountMap = new EnumMap<>(RealmCacheType.class);
+        for (RealmCacheType type : RealmCacheType.values()) {
+            refAndCountMap.put(type, new RefAndCount());
+        }
+    }
+
+    /**
+     * Creates a new Realm instance or get a existing instance for current thread.
+     *
+     * @param configuration {@link RealmConfiguration} will be used to create or get the instance.
+     * @param realmClass class of {@link Realm} or {@link DynamicRealm} to be created in or gotten from the cache.
+     * @return the {@link Realm} or {@link DynamicRealm} instance.
+     */
+    @SuppressWarnings("unchecked")
+    static synchronized <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
+                                                        Class<E> realmClass) {
+        boolean isCacheInMap = true;
+        RealmCache cache = cachesMap.get(configuration.getPath());
+        if (cache == null) {
+            // Create a new cache
+            cache = new RealmCache(configuration);
+            // The new cache should be added to the map later
+            isCacheInMap = false;
+        } else {
+            // != takes no time, check it first
+            if (cache.configuration != configuration && cache.configuration.hashCode() != configuration.hashCode()) {
+                // throw the exception
+                validateAgainstExistingConfiguration(configuration, cache.configuration);
+            }
+        }
+
+        RefAndCount refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+
+        if (refAndCount.localRealm.get() == null) {
+            // Create a new local Realm instance
+            BaseRealm realm;
+
+            if (realmClass == Realm.class) {
+                // RealmMigrationNeededException might be thrown here
+                realm = Realm.createInstance(configuration, cache.typedColumnIndices);
+            } else {
```

Comment:
Exception will be thrown if it is BaseRealm in line 108. But i will add the `if else` case to make it clearer.